File: browser/host.js
// Cairn's host interface, in JavaScript, around the engine the browser already has.
//
// # Why there is no WebAssembly engine in this directory
//
// There is a perfectly good one in the page, and using it is the entire point of
// [ADR-0005](../docs/adr/0005-the-fast-path-cannot-snapshot.md). A volunteer executes the
// canonical module on V8 or SpiderMonkey at full speed and returns a result. It cannot produce
// a trace — no browser exposes the operand stack, and none ever will — so it does not try. If
// somebody disputes the result, a *different* path re-executes the same unit under Cairn's
// interpreter, and that path is Rust and lives in `runtime/`.
//
// So this file is glue, and being glue is the achievement. Three imported functions and a
// global to read afterwards is the whole contract between the network and a volunteer.
//
// # What it does not do
//
// It does not instrument. The bytes it runs are the canonical binary a coordinator produced at
// registration — `cairn-worker prepare` writes exactly those — and their hash is the unit's
// identity. A volunteer that could rewrite its own work unit would be a volunteer whose result
// means nothing.

/// A workload cannot be trusted with the length it passes to `output`.
///
/// This is the same hazard the Rust hosts guard, and it is worth guarding in the same order
/// for the same reason: bounds-check *before* allocating. `len` is a value the workload chose,
/// a generated one will happily ask for four gigabytes, and a copy that large is a denial of
/// service against the volunteer's own machine dressed up as a result.
function readMemory(memory, ptr, len) {
  const start = ptr >>> 0;
  const count = len >>> 0;
  if (start + count > memory.buffer.byteLength) {
    return null;
  }
  return new Uint8Array(memory.buffer, start, count).slice();
}

/**
 * Execute one work unit and report what it produced.
 *
 * @param {BufferSource} bytes  The canonical module, as `cairn-worker prepare` wrote it.
 * @param {Uint8Array}   input  The unit's input.
 * @returns {Promise<{output: Uint8Array, fuel: bigint|null, milliseconds: number}>}
 */
export async function runUnit(bytes, input) {
  let memory = null;
  let output = new Uint8Array(0);

  // Counted only when the module meters through the host call. The honest path does not meter
  // at all and the global encoding does not call this, so on the two configurations a
  // volunteer actually receives, this stays zero — but the import must exist regardless,
  // because the instrumentation pass appends it to *every* module it emits, whether or not
  // anything calls it. Leaving it out fails instantiation, which is a confusing way to
  // discover that the pass keeps function indices stable by always adding the import.
  let chargedFuel = 0n;

  const imports = {
    cairn: {
      // Copies up to `len` bytes to `ptr` and returns the input's true length, so a workload
      // can size its buffer with a zero-length probe. The example unit does exactly that.
      input: (ptr, len) => {
        const available = input.length;
        const count = Math.min(available, len >>> 0);
        if (count > 0 && memory) {
          const start = ptr >>> 0;
          if (start + count <= memory.buffer.byteLength) {
            new Uint8Array(memory.buffer).set(input.subarray(0, count), start);
          }
          // A failed write means the workload named an out-of-bounds address. It will trap on
          // its own shortly; the host does not need to judge.
        }
        return available;
      },

      output: (ptr, len) => {
        if (!memory) return;
        const bytes = readMemory(memory, ptr, len);
        if (bytes) output = bytes;
      },

      charge: (instructions) => {
        chargedFuel += BigInt(instructions >>> 0);
      },
    },
  };

  const { instance } = await WebAssembly.instantiate(bytes, imports);
  memory = instance.exports.memory;

  const started = performance.now();
  instance.exports.cairn_run();
  const milliseconds = performance.now() - started;

  return { output, fuel: readFuel(instance, chargedFuel), milliseconds };
}

/// How much work the unit was, if the unit was prepared to say.
///
/// The interesting half of [ADR-0009](../docs/adr/0009-metering-through-a-global-the-engines-disagree.md):
/// a module metered through an exported counter can be run by an engine Cairn does not control
/// and still report an exact, machine-independent instruction count. Two lines here; it was
/// unavailable at any acceptable price when metering meant a host call per basic block.
///
/// `null` means the unit was not prepared with `--count-fuel`, which is the default. There is
/// no way to recover the number after the fact, and no reason to want one — the volunteer did
/// the work either way.
function readFuel(instance, chargedFuel) {
  const counter = instance.exports.cairn_fuel;
  if (counter) return counter.value;
  return chargedFuel > 0n ? chargedFuel : null;
}

File: worker-native/src/host.rs
//! The host side of the Cairn interface, for a volunteer's own engine.
//!
//! Three imports and nothing else. A workload can read its input, write its result, and — on
//! the dispute path only — report how far it has executed. It cannot reach a clock, entropy,
//! the filesystem or the network, because those imports are not offered and
//! [`cairn_runtime::validate`] rejects a module that asks for anything else.
//!
//! `cairn.charge` is linked here even though the honest path never calls it, so that the same
//! host works for a module instrumented either way. On the honest path the counter simply stays
//! at zero.

use cairn_runtime::validate;
use wasmtime::{Caller, Engine, Extern, Linker, Module, Store};

/// What the host holds on a workload's behalf.
#[derive(Default)]
struct Host {
    input: Vec<u8>,
    output: Vec<u8>,
    fuel: u64,
}

/// What one execution produced.
pub struct Executed {
    /// The bytes the workload wrote through `cairn.output`.
    pub output: Vec<u8>,
    /// How much work it was, if the unit was prepared to say.
    ///
    /// The interesting half of [ADR-0009]: a module metered through an exported counter runs on
    /// an engine Cairn does not control and still reports an exact, machine-independent
    /// instruction count. `None` means the module was not prepared to count, which is the
    /// default — there is no way to recover the number afterwards and no reason to want one.
    ///
    /// [ADR-0009]: ../../docs/adr/0009-metering-through-a-global-the-engines-disagree.md
    pub fuel: Option<u64>,
}

/// Compile and run `module` on wasmtime, returning what it wrote through `cairn.output`.
///
/// # Errors
///
/// Any failure to compile, link, instantiate or execute, as a message fit to print.
pub fn execute(module: &[u8], input: &[u8]) -> Result<Executed, String> {
    let engine = Engine::default();
    let module = Module::new(&engine, module).map_err(|e| format!("could not compile: {e}"))?;
    let mut store = Store::new(
        &engine,
        Host {
            input: input.to_vec(),
            ..Host::default()
        },
    );
    let mut linker = <Linker<Host>>::new(&engine);

    linker
        .func_wrap(
            validate::HOST_MODULE,
            validate::HOST_CHARGE,
            |mut caller: Caller<'_, Host>, instructions: i32| {
                caller.data_mut().fuel += u64::from(instructions as u32);
            },
        )
        .map_err(|e| format!("could not link charge: {e}"))?;

    linker
        .func_wrap(
            validate::HOST_MODULE,
            validate::HOST_INPUT,
            |mut caller: Caller<'_, Host>, ptr: i32, len: i32| -> i32 {
                // Returns the full length available, whatever was asked for, so a workload can
                // size its buffer by calling with a length of zero first.
                let available = caller.data().input.len();
                let count = available.min(len as u32 as usize);
                if count > 0 {
                    let bytes = caller
                        .data()
                        .input
                        .get(..count)
                        .unwrap_or_default()
                        .to_vec();
                    if let Some(memory) = caller.get_export("memory").and_then(Extern::into_memory)
                    {
                        // A failed write means the workload named an out-of-bounds address. It
                        // will trap on its own shortly; the host does not need to judge.
                        let _ = memory.write(&mut caller, ptr as u32 as usize, &bytes);
                    }
                }
                available as i32
            },
        )
        .map_err(|e| format!("could not link input: {e}"))?;

    linker
        .func_wrap(
            validate::HOST_MODULE,
            validate::HOST_OUTPUT,
            |mut caller: Caller<'_, Host>, ptr: i32, len: i32| {
                let Some(memory) = caller.get_export("memory").and_then(Extern::into_memory) else {
                    return;
                };
                // Bounds-check *before* allocating. `len` is chosen by the workload, and this
                // host runs workloads a volunteer did not write: `vec![0u8; len]` on a length
                // of `0xffff_ffff` asks the operating system for four gigabytes, which is a
                // denial of service against the volunteer's own machine dressed up as a result.
                // The read itself would fail, but only after the allocation succeeded or the
                // process died trying.
                let start = ptr as u32 as usize;
                let count = len as u32 as usize;
                if start
                    .checked_add(count)
                    .is_none_or(|end| end > memory.data_size(&caller))
                {
                    return;
                }
                let mut buffer = vec![0u8; count];
                if memory.read(&caller, start, &mut buffer).is_ok() {
                    caller.data_mut().output = buffer;
                }
            },
        )
        .map_err(|e| format!("could not link output: {e}"))?;

    let instance = linker
        .instantiate(&mut store, &module)
        .map_err(|e| format!("could not instantiate: {e}"))?;
    let entry = instance
        .get_typed_func::<(), ()>(&mut store, validate::ENTRY_POINT)
        .map_err(|e| format!("module does not export {}: {e}", validate::ENTRY_POINT))?;

    entry
        .call(&mut store, ())
        .map_err(|e| format!("execution trapped: {e}"))?;

    // Two encodings, one number. `Metering::Global` counts into an exported global — free enough
    // to leave on — and `Metering::HostCall` counts through the import linked above. A module
    // metered neither way reports nothing, which is not a failure.
    let counted = instance
        .get_global(&mut store, validate::FUEL_EXPORT)
        .and_then(|global| global.get(&mut store).i64())
        .map(|value| value as u64);
    let charged = store.data().fuel;

    Ok(Executed {
        output: std::mem::take(&mut store.data_mut().output),
        fuel: counted.or((charged > 0).then_some(charged)),
    })
}
